### ISSUE TITLE
"Preparing build directory" step should also (probably) prune remote branches

### ISSUE THREAD

**Issue by SEAPUNK** [buildkite/agent#507]
Sometimes, if I create a branch called `foo/bar/baz`, where branch `foo/bar` exists, I cannot push the branch to the remote Github repo. So, because `foo/bar` on the Github repo isn't used, I delete that branch, run `git remote prune origin`, and then push my branch. This causes a problem running `git fetch` on the build box:

```
  | $ git clean -fdq
  | $ git submodule foreach --recursive git clean -fdq
  | $ git fetch -v origin aaaaaaaaa442166ea29ce572f4cbc6385b5a9266
  | $ git fetch -v origin +refs/heads/\*:refs/remotes/origin/\* +refs/tags/\*:refs/tags/\*
  | remote: Counting objects: 25, done.
  | remote: Compressing objects: 100% (5/5), done.
  | remote: Total 25 (delta 21), reused 24 (delta 20), pack-reused 0
  | Unpacking objects: 100% (25/25), done.
  | error: 'refs/remotes/origin/foo/bar' exists; cannot create 'refs/remotes/origin/foo/bar/baz'
  | From github.com:some/repo
  | ! [new branch]      foo/bar/baz -> origin/foo/bar/baz  (unable to update local ref)
```

causing all of the subsequent builds to fail.

Because of this, to resolve this issue, maybe the agent should run `git remote prune origin` shortly after running `git clean`, to prevent this error from occurring?

**Comment by cysp** [comment]
Alternatively, `git fetch` takes a `--prune` option which could combine those steps.

**Comment by lox** [comment]
Closed in #518
